src/scatter.jl

```diff
@@ -0,0 +1,94 @@
+export scatter!
+
+"""
+    scatter!(op, dst, src, idx, dims)
+
+Scatter operation, which scatters data in `src` and assigns to `dst` according to `idx`.
+With the data going to the same place, specified operation is applied on to reduce data.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(dst[idx[k]...], src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `dst`: the destination for `src` to aggregate to. This argument will be mutated.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+- `dims`: the number of dimensions to be view as an operational unit from the beginning.
+If dims=0 is given, operational unit is scalar. Default dims is 1.
+
+Examples for dims are lists here:
+
+- `dims=0`: dst[idx[k]...] = op(dst[idx[k]...], src[k...])
+- `dims=1`: dst[:, idx[k]...] = (op).(dst[:, idx[k]...], src[k...])
+- `dims=2`: dst[:, :, idx[k]...] = (op).(dst[:, :, idx[k]...], src[k...])
+"""
+function scatter!(op, dst::AbstractArray{T}, src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple},
+                  dims::Integer=1) where {T<:Real}
+    if dims > 0
+        scatter_vec!(op, dst, src, idx, dims)
```

Comment:
```suggestion
        scatter_vec!(op, dst, src, idx, Val(dims))
```
Since `dims` affects the types, maybe it should be passed as Val.  Then `scatter_vec!` which actually does the work can be type-stable.

Reply:
Note that `scatter_vec!` will also have to accept the argument like `, ::Val{dim}) where {dim}` instead of `, dim::Int)` for this to work.